core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1869,12 +1869,16 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def foreachParN[R >: LowerR, E <: UpperE, A, B](
     n: Long
   )(as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    (for {
-      semaphore <- Semaphore.make(n)
-      bs <- foreachPar[R, E, A, B](as) { a =>
-             semaphore.withPermit(fn(a))
-           }
-    } yield bs).refailWithTrace
+    for {
+      q     <- Queue.bounded[(Promise[E, B], A)](n.toInt)
+      env   <- ZIO.environment[R]
+      pairs <- ZIO.foreach(as)(a => Promise.make[E, B].map(p => (p, a)))
+      _     <- ZIO.foreach(pairs)(pair => q.offer(pair)).fork
+      _ <- ZIO.collectAll(List.fill(n.toInt)(q.take.flatMap {
+            case (p, a) => p.done(fn(a).provide(env).onError(_ => q.shutdown))
```

Comment:
After some experimenting I realized that p.done(fn(a)) implies that the effect produced by fn(a) will be evaluated by the fiber that invokes p.await. In this case that is the fiber that sequentially awaits all the promises.

In practice this means we don't have any parallelism with the proposed implementation.

A reasonable fix is to move the fn(a) call outside of p.done. Something like

fn(a).onError( _ => q.shutdown) >>= (p.succeed)

Might be a good idea to verify manually that the behavior is parallel as well. I can post some code later for that (away from my computer at the moment)

Reply:
You're totally right! Good catch! I just made the change you suggested and manually confirmed that effects are now running in parallel. I also added a test for this.

Reply:
Another observation is that 

`p.done(fn(a).provide(env).onError(_ => q.shutdown))`

doesn't propagate how fn(a) failed. I think a better way of doing this is

`fn(a).foldM(e => p.fail(e) *> q.shutdown, b => p.succeed(b)).onError(_ => q.shutdown`

That should propagate the error correctly (though some tests should be added to verify this as well)

Reply:
Best use `foldCauseM` and `p.halt` for that.

Reply:
I fixed this so that errors are propagated and added a test for it. Looks like everything is working now except one of the CI tests for `RTSSpec` is failing. I'm not sure what is going on here. It does not appear to be related to any of the changes in this PR.

Reply:
When using foldCauseM you don't need the onError.

Reply:
Got it. It seems using a collection size of 1,000,000 elements in the tests was causing some flakiness (at different times we were failing on 2.13 / JDK 8, 2.13 / JDK 11, and 2.12 / JDK 8 in what appeared to be unrelated tests) so I reduced it to 100,000 and everything is working now. What do you think?

Reply:
If you compare it to the original implementation that's still a big improvement. Sounds more than fair to me.